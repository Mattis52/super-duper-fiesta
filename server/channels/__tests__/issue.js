jest.mock('../../models/issue');
jest.mock('../../models/meeting');
jest.mock('../../models/user');
jest.mock('../../models/vote');
jest.mock('../../utils');
const { createIssue, closeIssue, adminDeleteIssue } = require('../issue');
const { addIssue, endIssue, deleteIssue, getActiveQuestion } = require('../../models/issue');
const { getActiveGenfors, getGenfors } = require('../../models/meeting');
const { getQualifiedUsers } = require('../../models/user');
const { getVotes } = require('../../models/vote');
const { emit, broadcast } = require('../../utils');
const { generateSocket, generateIssue, generateGenfors, generateUser, generateVote } = require('../../utils/generateTestData');

beforeEach(() => {
  addIssue.mockImplementation(async () => generateIssue());
  getActiveQuestion.mockImplementation(async () => null);
  endIssue.mockImplementation(async () => generateIssue());
  getActiveGenfors.mockImplementation(async () => generateGenfors({ id: '1' }));
  getGenfors.mockImplementation(async () => generateGenfors({ id: '1' }));
  getQualifiedUsers.mockImplementation(async () => [
    generateUser({ id: '2' }),
    generateUser({ id: '3' }),
    generateUser({ id: '4' }),
  ]);
  getVotes.mockImplementation(async ({ _id: issueId }) => [
    generateVote({ question: issueId, _id: '1' }),
    generateVote({ question: issueId, _id: '2' }),
    generateVote({ question: issueId, _id: '3' }),
    generateVote({ question: issueId, _id: '4' }),
  ]);
});

describe('createIssue', () => {
  const generateData = () => ({});
  it('emits OPEN_ISSUE action creates issue', async () => {
    await createIssue(generateSocket(), generateData());

    expect(emit.mock.calls).toMatchSnapshot();
    expect(broadcast.mock.calls).toMatchSnapshot();
  });

  it('emits an error when it fails', async () => {
    addIssue.mockImplementation(async () => { throw new Error('Failed'); });
    await createIssue(generateSocket(), generateData());

    expect(emit.mock.calls).toMatchSnapshot();
    expect(broadcast.mock.calls).toEqual([]);
  });
});

describe('closeIssue', () => {
  const generateData = () => ({
    issue: generateIssue(),
  });

  it('emits close issue action and disables voting', async () => {
    endIssue.mockImplementation(async () => generateIssue({ active: false }));
    await closeIssue(generateSocket({ permissions: 10 }), generateData());

    expect(emit.mock.calls).toMatchSnapshot();
    expect(broadcast.mock.calls).toMatchSnapshot();
  });

  it('emits error when it fails', async () => {
    endIssue.mockImplementation(async () => { throw new Error('Failed'); });
    await closeIssue(generateSocket(), generateData());

    expect(emit.mock.calls).toMatchSnapshot();
    expect(broadcast.mock.calls).toEqual([]);
  });
});

describe('adminDeleteIssue', () => {
  const generateData = () => ({
    issue: generateIssue(),
  });
  it('emits delete issue on success', async () => {
    deleteIssue.mockImplementation(async () => generateIssue({ active: false, deleted: true }));

    await adminDeleteIssue(generateSocket({ permissions: 10 }), generateData());

    expect(emit.mock.calls).toMatchSnapshot();
    expect(broadcast.mock.calls).toMatchSnapshot();
  });
});
