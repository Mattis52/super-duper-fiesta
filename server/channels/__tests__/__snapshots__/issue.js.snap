// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adminDeleteIssue emits delete issue on success 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "hashy",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": true,
          "genfors": "1",
          "logged_in": true,
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 10,
        },
      },
    },
    "CLOSE_ISSUE",
    Object {
      "_id": "1",
      "active": false,
      "countingBlankVotes": false,
      "deleted": true,
      "description": "Description goes here",
      "genfors": "1",
      "options": Array [
        Object {
          "id": "1",
          "text": "Blank",
        },
        Object {
          "id": "2",
          "text": "Yes",
        },
        Object {
          "id": "3",
          "text": "No",
        },
      ],
      "secret": false,
      "showOnlyWinner": false,
      "toObject": [Function],
      "voteDemand": "regular",
      "votes": Object {
        "1": Object {
          "_id": "1",
          "option": "3",
          "question": "1",
          "user": "1",
        },
        "2": Object {
          "_id": "2",
          "option": "3",
          "question": "1",
          "user": "1",
        },
        "3": Object {
          "_id": "3",
          "option": "3",
          "question": "1",
          "user": "1",
        },
        "4": Object {
          "_id": "4",
          "option": "3",
          "question": "1",
          "user": "1",
        },
      },
      "winner": "3",
    },
  ],
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "hashy",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": true,
          "genfors": "1",
          "logged_in": true,
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 10,
        },
      },
    },
    "DELETED_ISSUE",
    Object {
      "_id": "1",
      "active": false,
      "countingBlankVotes": false,
      "deleted": true,
      "description": "Description goes here",
      "genfors": "1",
      "options": Array [
        Object {
          "id": "1",
          "text": "Blank",
        },
        Object {
          "id": "2",
          "text": "Yes",
        },
        Object {
          "id": "3",
          "text": "No",
        },
      ],
      "secret": false,
      "showOnlyWinner": false,
      "toObject": [Function],
      "voteDemand": "regular",
      "votes": Object {
        "1": Object {
          "_id": "1",
          "option": "3",
          "question": "1",
          "user": "1",
        },
        "2": Object {
          "_id": "2",
          "option": "3",
          "question": "1",
          "user": "1",
        },
        "3": Object {
          "_id": "3",
          "option": "3",
          "question": "1",
          "user": "1",
        },
        "4": Object {
          "_id": "4",
          "option": "3",
          "question": "1",
          "user": "1",
        },
      },
      "winner": "3",
    },
  ],
]
`;

exports[`adminDeleteIssue emits delete issue on success 2`] = `
Array [
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "hashy",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": true,
          "genfors": "1",
          "logged_in": true,
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 10,
        },
      },
    },
    "DISABLE_VOTING",
  ],
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "hashy",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": true,
          "genfors": "1",
          "logged_in": true,
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 10,
        },
      },
    },
    "CLOSE_ISSUE",
    Object {
      "_id": "1",
      "active": false,
      "countingBlankVotes": false,
      "deleted": true,
      "description": "Description goes here",
      "genfors": "1",
      "options": Array [
        Object {
          "id": "1",
          "text": "Blank",
        },
        Object {
          "id": "2",
          "text": "Yes",
        },
        Object {
          "id": "3",
          "text": "No",
        },
      ],
      "secret": false,
      "showOnlyWinner": false,
      "toObject": [Function],
      "voteDemand": "regular",
    },
  ],
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "hashy",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": true,
          "genfors": "1",
          "logged_in": true,
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 10,
        },
      },
    },
    "DELETED_ISSUE",
    Object {
      "_id": "1",
      "active": false,
      "countingBlankVotes": false,
      "deleted": true,
      "description": "Description goes here",
      "genfors": "1",
      "options": Array [
        Object {
          "id": "1",
          "text": "Blank",
        },
        Object {
          "id": "2",
          "text": "Yes",
        },
        Object {
          "id": "3",
          "text": "No",
        },
      ],
      "secret": false,
      "showOnlyWinner": false,
      "toObject": [Function],
      "voteDemand": "regular",
      "votes": Object {
        "1": Object {
          "_id": "1",
          "option": "3",
          "question": "1",
          "user": "1",
        },
        "2": Object {
          "_id": "2",
          "option": "3",
          "question": "1",
          "user": "1",
        },
        "3": Object {
          "_id": "3",
          "option": "3",
          "question": "1",
          "user": "1",
        },
        "4": Object {
          "_id": "4",
          "option": "3",
          "question": "1",
          "user": "1",
        },
      },
      "winner": "3",
    },
  ],
]
`;

exports[`closeIssue emits close issue action and disables voting 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "hashy",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": true,
          "genfors": "1",
          "logged_in": true,
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 10,
        },
      },
    },
    "CLOSE_ISSUE",
    Object {
      "_id": "1",
      "active": false,
      "countingBlankVotes": false,
      "description": "Description goes here",
      "genfors": "1",
      "options": Array [
        Object {
          "id": "1",
          "text": "Blank",
        },
        Object {
          "id": "2",
          "text": "Yes",
        },
        Object {
          "id": "3",
          "text": "No",
        },
      ],
      "secret": false,
      "showOnlyWinner": false,
      "toObject": [Function],
      "voteDemand": "regular",
      "votes": Object {
        "1": Object {
          "_id": "1",
          "option": "3",
          "question": "1",
          "user": "1",
        },
        "2": Object {
          "_id": "2",
          "option": "3",
          "question": "1",
          "user": "1",
        },
        "3": Object {
          "_id": "3",
          "option": "3",
          "question": "1",
          "user": "1",
        },
        "4": Object {
          "_id": "4",
          "option": "3",
          "question": "1",
          "user": "1",
        },
      },
      "winner": "3",
    },
  ],
]
`;

exports[`closeIssue emits close issue action and disables voting 2`] = `
Array [
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "hashy",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": true,
          "genfors": "1",
          "logged_in": true,
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 10,
        },
      },
    },
    "DISABLE_VOTING",
  ],
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "hashy",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": true,
          "genfors": "1",
          "logged_in": true,
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 10,
        },
      },
    },
    "CLOSE_ISSUE",
    Object {
      "_id": "1",
      "active": false,
      "countingBlankVotes": false,
      "description": "Description goes here",
      "genfors": "1",
      "options": Array [
        Object {
          "id": "1",
          "text": "Blank",
        },
        Object {
          "id": "2",
          "text": "Yes",
        },
        Object {
          "id": "3",
          "text": "No",
        },
      ],
      "secret": false,
      "showOnlyWinner": false,
      "toObject": [Function],
      "voteDemand": "regular",
      "votes": Object {
        "1": Object {
          "_id": "1",
          "option": "3",
          "question": "1",
          "user": "1",
        },
        "2": Object {
          "_id": "2",
          "option": "3",
          "question": "1",
          "user": "1",
        },
        "3": Object {
          "_id": "3",
          "option": "3",
          "question": "1",
          "user": "1",
        },
        "4": Object {
          "_id": "4",
          "option": "3",
          "question": "1",
          "user": "1",
        },
      },
      "winner": "3",
    },
  ],
]
`;

exports[`closeIssue emits error when it fails 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "hashy",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": true,
          "genfors": "1",
          "logged_in": true,
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 5,
        },
      },
    },
    "issue",
    Object {},
    Object {
      "error": "Closing issue failed",
    },
  ],
]
`;

exports[`createIssue emits OPEN_ISSUE action creates issue 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "hashy",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": true,
          "genfors": "1",
          "logged_in": true,
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 5,
        },
      },
    },
    "OPEN_ISSUE",
    Object {
      "_id": "1",
      "active": true,
      "countingBlankVotes": false,
      "description": "Description goes here",
      "genfors": "1",
      "options": Array [
        Object {
          "id": "1",
          "text": "Blank",
        },
        Object {
          "id": "2",
          "text": "Yes",
        },
        Object {
          "id": "3",
          "text": "No",
        },
      ],
      "secret": false,
      "showOnlyWinner": false,
      "toObject": [Function],
      "voteDemand": "regular",
    },
    Object {
      "action": "open",
    },
  ],
]
`;

exports[`createIssue emits OPEN_ISSUE action creates issue 2`] = `
Array [
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "hashy",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": true,
          "genfors": "1",
          "logged_in": true,
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 5,
        },
      },
    },
    "OPEN_ISSUE",
    Object {
      "_id": "1",
      "active": true,
      "countingBlankVotes": false,
      "description": "Description goes here",
      "genfors": "1",
      "options": Array [
        Object {
          "id": "1",
          "text": "Blank",
        },
        Object {
          "id": "2",
          "text": "Yes",
        },
        Object {
          "id": "3",
          "text": "No",
        },
      ],
      "secret": false,
      "showOnlyWinner": false,
      "toObject": [Function],
      "voteDemand": "regular",
    },
    Object {
      "action": "open",
    },
  ],
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "hashy",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": true,
          "genfors": "1",
          "logged_in": true,
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 5,
        },
      },
    },
    "ENABLE_VOTING",
  ],
]
`;

exports[`createIssue emits an error when it fails 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "hashy",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": true,
          "genfors": "1",
          "logged_in": true,
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 5,
        },
      },
    },
    "issue",
    Object {},
    Object {
      "error": "Adding new question failed",
    },
  ],
]
`;
